### ISSUE TITLE
New rule: `confusing-non-null-assertion-like-not-equal`

### ISSUE THREAD

**Issue by zhangciwu** [typescript-eslint/typescript-eslint#1940]
Using a non-null assertions(`!`) and an equals test(`==`) together makes code looks very similar with not equals test(`!==`), look at below code:

```typescript
a! == b; // a non-null assertions(`!`) and an equals test(`==`)
a !== b; // not equals test(`!==`)
a! === b; // a non-null assertions(`!`) and an triple equals test(`===`)
```

I have submit an issue on typescript repo [`Issue: Easy misunderstanding: "! ==="`](https://github.com/microsoft/TypeScript/issues/37837), and a PR to fix it.

And I think it's good for many people to avoid this writing by composing a lint rule

**Comment by bradzacher** [comment]
though I wouldn't agree that `a! === b` is confusing (because `!===` is not valid code), I do agree that `a! == b` is both redundant and confusing - there's no need for a non-null assertion with an equality check.

**Comment by zhangciwu** [comment]
> though I wouldn't agree that `a! === b` is confusing (because `!===` is not valid code), I do agree that `a! == b` is both redundant and confusing - there's no need for a non-null assertion with an equality check.

For human eyes and brain, it could be easy to combine `!` and `=`s into `not eqauls`, especially at fast code scaning within a bunch of code; it feels like, first it's `not equals`, and then hold on, somthing weird, then realise it's three equals sign, and after some time, figure out it's `non-null assertion` 
(Basiclly it's my brain activities when I first see such code in typescript repo)

So `!===` is not valid code, but not that obvious for some people

**Comment by bbarry** [comment]
Is there ever a reason to use a non-null assertion in any comparison?

I'd suggest a rule `no-unnecessary-non-null-assertion` (haven't looked to see if there is one already) that forbids the non-null assertion operator in any comparison (or even any context that permits null?).

The following should all be invalid (not exhaustive):

    !!a!

    a! === b

    a! !== b

    a! < b

    function foo(a: number | null) { }
    foo(a!);


The following should be valid:

    a!.b;

    let a: {
        s?(): number,
        b?: number[]
    } = {};
    a.s!();
    a.b![0];

    let a: {
        s: (()=> number) | null,
        b: number[] | null
    } = { s:null, b: null};
    a.s!();
    a.b![0];


    function foo(a: number) { }
    foo(a!);

**Comment by zhangciwu** [comment]
@bbarry  exporession like below not include:
```
let a: number = 1;
let b: {x?: number} = {};
let result = a + b.x! == c;
```

And for function parameters

**Comment by bradzacher** [comment]
That's a good point @bbarry - thanks.
This would definitely be a good idea as an extension to `no-unnecessary-non-null-assertion`.

We can easily analyse the AST to detect the case where there is a non-null assertion on the left or right of a `==`/`!=`/`===`/`!==` comparison.

The only case we'd be missing is `a + b.x! == c`.

Note that:
- `a! < b` this is a valid non-null assertion, as if `a: number | null`, then `a < b` is a type error, as TS does not allow `<` comparisons on a `null`.
  - it is also not confusing, as `!<` is not an operator.
- `function foo(a: number | null) { }   foo(a!);` is handled by `no-unnecessary-type-assertion`.

**Comment by zhangciwu** [comment]
@bbarry  It's good, and for a more thorough coverage, this could be also bad case:
```
if(a.x){
    a.x!
}
```

**Comment by zhangciwu** [comment]
There is a rule named `no-extra-non-null-assertion.ts` [doc link](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-extra-non-null-assertion.md), covers some scenario that @bbarry mentioned, 
but still others not covered, mostly contextual positions that allows `null|undefined` (like function params)

Is there a way to retrive the required type of a position of AST node? like param type in function invocation
